docs/dev-process/coding-guidelines.md

```diff
@@ -1,92 +1,179 @@
-* File encoding should be ASCII (preferred)
-  or UTF8 (with BOM) if absolutely necessary.
-
-## Preprocessor defines
 
-There are 3 primary preprocessor macros we define during builds:
+### Commenting Conventions
 
-* DEBUG - guard code that should not be included in release builds
-* CORECLR - guard code that differs between Full CLR and CoreCLR
-* UNIX - guard code that is specific to Unix (Linux and macOS)
+* Add comments when changes are not trivial or could be confusing.
 
-Any other preprocessor defines found in the source are used for one-off custom builds,
-typically to help debug specific scenarios.
+* Add comments when a reviewer needs help to understand your changes.
 
-### Runtimes
+* Update/remove existing comments when you are changing the corresponding code.
 
-The PowerShell repo is used to build PowerShell targeting CoreCLR as well as CLR 4.5.
+* Make sure the added/updated comments are accurate and easy to understand.
 
-Code under !CORECLR must build against CLR 4.5.
-We will not accept changes that require a later version of the full CLR.
-In extremely rare cases, we may use reflection to use an API in a later version of the CLR,
-but the feature must robustly handle running with CLR 4.5.
+* Public members must use [doc comments](https://msdn.microsoft.com/en-us/library/b2s063f7.aspx).
+  Internal and private members may use doc comments but it is not required.
 
-We may reject code under !CORECLR without explanation because
-we do not support installation or testing of such code in this repo.
-All new features should support CoreCLR.
-
-## Performance considerations
+## Performance Considerations
 
 PowerShell has a lot of performance sensitive code as well as a lot of inefficient code.
-We have some guidelines that we typically apply widely even in less important code
-because code and patterns are copied we want certain inefficient code to stay out of the performance critical code.
+We have some guidelines that we typically apply widely even in less important code because code and patterns are copied,
+and we want certain inefficient code to stay out of the performance critical code.
 
 Some general guidelines:
 
-* Avoid LINQ - it can create lots of avoidable garbage
+* Avoid LINQ - it can create lots of avoidable garbage.
+
 * Prefer `for` and `foreach`,
   with a slight preference towards `for` when you're uncertain if `foreach` allocates an iterator.
+
 * Avoid `params` arrays, prefer adding overloads with 1, 2, 3, and maybe more parameters.
+
 * Be aware of APIs such as `String.Split(params char[])` that do not provide overloads to avoid array allocation.
-  When calling such APIs, reuse a static array when possible.
+  When calling such APIs, reuse a static array when possible (i.e. `Utils.Separators.Colon`).
+
+* Avoid creating empty arrays.
+  Instead, reuse the static ones via `Utils.EmptyArray<T>`.
+
 * Avoid unnecessary memory allocation in a loop.
   Move the memory allocation outside the loop if possible.
```

Comment:
I remember a recommendation to use static RegEx to benefit caching. Is this relevant? Maybe still mention compile RegEx and using StringBuilder if we have more 3-4 re-allocations.

Reply:
I think StringBuilder is worth mentioning -- string operations are quite common. `RegEx` is too specific to me.